### PR TITLE
core/src: Add half single and double mixed compare (LT,GT,LE,GE)

### PR DESCRIPTION
**Pull request by e10harvey**
Related to #6404:
- Support mixed precision in operator>, operator>=, operator<=
- Do not downcast `T` for operator<

### CODE REVIEW

**Comment by e10harvey** [comment]
@dalg24, @crtrott: Shall I add all mixed precision comparison operators to this PR or would a separate issue and PR be preffered?

**Comment by aprokop** [comment]
Should the tests be expanded to include bfloat16?

**Comment by e10harvey** [comment]
> Should the tests be expanded to include bfloat16?

Yes, bfloat16 is covered by these tests.

**Comment by e10harvey** [comment]
retest this please

**Comment by crtrott** [comment]
Does this support half_t vs bhalf_t comparison? Also what about `1<half_t(2)` that doesn't work or? I wonder if we should fix that. For example we could have comparisons explicitly for float and double (not as templated functions) and then have a templated one which converts both args to double or to T?

**Comment by e10harvey** [comment]
> Does this support half_t vs bhalf_t comparison?
No.

> Also what about `1<half_t(2)` that doesn't work or? I wonder if we should fix that. For example we could have comparisons explicitly for float and double (not as templated functions) and then have a templated one which converts both args to double or to T?
WRT `1<half_t(2)`, I can check this.

Shall I add all of these mixed precision comparison operators to this PR or would a separate issue and PR be preferred?

**Comment by crtrott** [comment]
I think it might be better to do it in one go because there could be issues of ambiguity? @dalg24 thoughts?

**Comment by crtrott** [comment]
Ok convinced now separate PRs are better.

**Comment by crtrott** [comment]
Oh but should we have `>=` `<=`

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
* What about operator== and operator!= ?
* Make sure your description clearly states that this PR is adding comparison operators when the other operand is float or double.
* Also I don't understand
  > Do not downcast `T` for operator<`

  in the description.  What do you mean?

**Comment by e10harvey** [comment]
> * What about operator== and operator!= ?

https://github.com/kokkos/kokkos/issues/6412

> * Make sure your description clearly states that this PR is adding comparison operators when the other operand is float or double.

Done.

> * Also I don't understand
>       > Do not downcast `T` for operator<`
>       
>       
>       in the description.  What do you mean?

See https://github.com/kokkos/kokkos/issues/6404#issuecomment-1706731457

**Comment by dalg24** [comment]
> > * What about operator== and operator!= ?
> 
> #6412

I don't understand your answer

**Comment by e10harvey** [comment]
> > > * What about operator== and operator!= ?
> > 
> > 
> > #6412
> 
> I don't understand your answer

Oh, we agreed to do that in a follow-on issue / PR during the developer meeting yesterday. This PR was just supposed to address the bug report in #6404 although I added >= and <= too. I filed #6412 to address adding the remaining operator overloads.

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 1]
```diff
@@ -820,27 +820,83 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs < tmp_rhs;
   }
 
+  template <class T>
```
Comment (nliber) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: Why template functions and not non-template overloads?

I'm asking because they don't do the same thing (such as how implicit conversions are handled).

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 2]
```diff
@@ -820,27 +820,83 @@ class alignas(FloatType) floating_point_wrapper {
...
+  KOKKOS_FUNCTION friend std::enable_if_t<
+      std::is_same<T, float>::value || std::is_same<T, double>::value, bool>
```
Comment (nliber) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:825]: Minor:  Use `is_same_v<...>` instead of is_same<...>::value`.

**Review by crtrott — APPROVED** [review]

**Review by nliber — CHANGES_REQUESTED** [review]

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: the template versions in principle allow us to group things together (here float and double), we can do the same thing for the integer versions in another PR where we would have to cast the half precision thing to float instead of T. From an implicit conversion thing for the common case this just changes from "no-overload found" to "ambiguous overload" or? Say we had a float and a double non-template overlaod and someone would compare with int?

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: > Why template functions and not non-template overloads?

I agree with @crtrott and would add that it reduces code duplication. But, I may be missing something. I do not see how this mixing member and non-member overloads of operatorOP would affect overload resolution. The compiler either:
1. has a conflict from the implicit conversion constructor and the member overload of operatorOP as well as the non-member overload for float and double or;
2. does not find a overload of operatorOP or;
3. finds the correct overload of operatorOP.

From https://en.cppreference.com/w/cpp/language/overload_resolution:

>  Call to an overloaded operator
>
> If at least one of the arguments to an operator in an expression has a class type or an enumeration type, both [builtin operators](https://en.cppreference.com/w/cpp/language/expressions#Operators) and [user-defined operator overloads](https://en.cppreference.com/w/cpp/language/operators) participate in overload resolution, with the set of candidate functions selected as follows:
>
> For a unary operator @ whose argument has type T1 (after removing cv-qualifications), or binary operator @ whose left operand has type T1 and right operand of type T2 (after removing cv-qualifications), the following sets of candidate functions are prepared:
> 1) member candidates: if T1 is a complete class or a class currently being defined, the set of member candidates is the result of [qualified name lookup](https://en.cppreference.com/w/cpp/language/lookup) of T1::operator@. In all other cases, the set of member candidates is empty.
> 2) non-member candidates: For the operators where [operator overloading](https://en.cppreference.com/w/cpp/language/operators) permits non-member forms, all declarations found by [unqualified name lookup](https://en.cppreference.com/w/cpp/language/lookup) of operator@ in the context of the expression (which may involve [ADL](https://en.cppreference.com/w/cpp/language/adl)), except that member function declarations are ignored and do not prevent the lookup from continuing into the next enclosing scope. If both operands of a binary operator or the only operand of a unary operator has enumeration type, the only functions from the lookup set that become non-member candidates are the ones whose parameter has that enumeration type (or reference to that enumeration type)
> 3) built-in candidates: For operator,, the unary operator&, and operator->, the set of built-in candidates is empty. For other operators built-in candidates are the ones listed in [built-in operator pages](https://en.cppreference.com/w/cpp/language/expressions#Operators) as long as all operands can be implicitly converted to their parameters. If any built-in candidate has the same parameter list as a non-member candidate or rewritten non-member candidate (since C++20) that is not a function template specialization, it is not added to the list of built-in candidates. When the built-in assignment operators are considered, the conversions from their left-hand arguments are restricted: user-defined conversions are not considered.
> 4) rewritten candidates:
>        For the four relational operator expressions x < y, x <= y, x > y, and x >= y, all member, non-member, and built-in operator<=>s found are added to the set.
 >       For the four relational operator expressions x < y, x <= y, x > y, and x >= y as well as the three-way comparison expression x <=> y, a synthesized candidate with the order of the two parameters reversed is added for each member, non-member, and built-in operator<=>s found.
 >       For x != y, all member, non-member, and built-in operator==s found are added to the set, unless there is a matching operator!=.
 >       For equality operator expressions x == y and x != y, a synthesized candidate with the order of the two parameters reversed is added for each member, non-member, and built-in operator==s found, unless there is a matching operator!=.
> In all cases, rewritten candidates are not considered in the context of the rewritten expression. For all other operators, the rewritten candidate set is empty. 
> ...
> User-defined conversions (both converting constructors and user-defined conversion functions) are prohibited from taking part in implicit conversion sequence where it would make it possible to apply more than one user-defined conversion. Specifically, they are not considered if the target of the conversion is the first parameter of a constructor or the implicit object parameter of a user-defined conversion function, and that constructor/user-defined conversion is a candidate for 
> ...
> Additional rules for function template candidates
> If name lookup produces a function template, [template argument deduction](https://en.cppreference.com/w/cpp/language/template_argument_deduction) and checking of any explicit template arguments are performed to find the template argument values (if any) that can be used in this case:
>    if both succeeds, the template arguments are used to synthesize declarations of the corresponding function template specializations, which are added to the candidate set, and such specializations are treated just like non-template functions except where specified otherwise in the tie-breaker rules below;
>    if argument deduction fails or the synthesized function template specialization would be ill-formed, no such function is added to the candidate set (see [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae)). 
> If a name refers to one or more function templates and also to a set of overloaded non-template functions, those functions and the specializations generated from the templates are all candidates.

Comment (nliber) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: Given

```c++
struct MyDouble {
    operator double() const { /* ... */ }
};
```

You could not pass `MyDouble` to it, so this doesn't act like built in comparison operators.

In this file, we have some comparisons that are member functions, some are non-templated `friend` functions, and now some that are templated `friend` functions, as well as sometimes passing by value to avoid `const`/`volatile` issues.  That combination is very hard to reason about.

Comment (nliber) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]:
```c++
struct MyDouble {
    operator double() const { /* ... */ }
};
```

You could not pass `MyDouble` to it, so this doesn't act like built in comparison operators.

In this file, we have some comparisons that are member functions, some are non-templated `friend` functions, and now some that are templated `friend` functions, as well as sometimes passing by value to avoid `const`/`volatile` issues.  That combination is very hard to reason about.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: @nliber I see. We can replace all the templated functions with non-templated functions. But, AFAIK, we cannot replace the non-member friend functions for all binary operators since they are used to overload the LHS type.

@crtrott, @dalg24, @ajpowelsnl: With half_t and bhalf_t in a minimally working state, I suggest we take a step back at this point and document a strategy for how to 'make half_t and bhalf_t behave like a c++ built-in type'. I am assuming this will not be easy given that the rules surrounding this are quite complicated.

Comment (nliber) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: @e10harvey the only operators that are required to be member functions are assignment `=`, subscript `[]`, call `()` and pointer indirection `->`.

For all the operators that don't take a non-const reference to the LHS, I'd rather they all be hidden `friend`s, which matches the current wisdom of the C++ standard library.  Newer libraries in the C++ standard library follow this convention.  See [P1601](https://wg21.link/P1601).

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: Got it. Thank you for the detailed review and guidance, @nliber.

Shall we make this change (everything but `=` operator implemented as a hidden friend) in this bug fix PR or in a follow-on issue/PR.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: CC: @crtrott, @dalg24

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: The failing HIP Jenkins CI SIGINT failure appears to be unrelated to these changes.

Comment (nliber) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: @e10harvey A follow-on is fine by me.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:823]: @nliber See https://github.com/kokkos/kokkos/issues/6468.

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 3]
```diff
@@ -820,27 +820,83 @@ class alignas(FloatType) floating_point_wrapper {
...
+  operator<(floating_point_wrapper lhs, T rhs) {
+    return T(lhs) < rhs;
+  }
+
+  template <class T>
+  KOKKOS_FUNCTION friend std::enable_if_t<
+      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
+  operator<(T lhs, floating_point_wrapper rhs) {
+    return lhs < T(rhs);
+  }
```
Comment (masterleinad) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: Wouldn't it be better to provide the overload only for `float` (and accept a possible narrowing conversion)? Types convertible to `float` should then still work (but don't with the current approach since both overloads would be feasible comparing, e.g., with an `int`).
 ```suggestion
  KOKKOS_FUNCTION friend bool
  operator<(floating_point_wrapper lhs, float rhs) {
    return T(lhs) < rhs;
  }
```

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: Note that the `SYCL` implementation just casts everything to `float`: https://github.com/intel/llvm/blob/83aafddac12e9b2df8e1764d3fba0ffe7ef6c596/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L176-L198.

**Review by nliber — APPROVED** [review]

**Review by masterleinad — CHANGES_REQUESTED** [review]
Please discuss https://github.com/kokkos/kokkos/pull/6407#pullrequestreview-1647070111.

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp (hunk 3 above)
Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: > Wouldn't it be better to provide the overload only for `double`? Types convertible to `double` should then still work (but don't with the current approach since both overloads would be feasible comparing, e.g., with an `int`

I like the simplicity of this approach. I believe that for other mixed precision expressions involving `T1 op T2`, the compiler will implicitly up-cast either `T1` or `T2`. So, providing only the double overload would result in unexpected behavior by up-casting half-precision operands to double. That is my understanding.

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: For a comparison with an `int`, https://godbolt.org/z/efnxf7e5n gives an error:
```
ambiguous overload for 'operator<' (operand types are 'Foo' and 'int')
```

If only providing the `double` or `float` overload we are getting a similar warning with GCC and an error with clang (if the conversion operator is not explicit).

If we use one templated version, everything works fine, and we can do whatever we want in the comparison operator. This does different casts depending on the types used, see https://cppinsights.io/s/2155d1dd.

If we don't provide the operator, implicit conversion kicks in which means we always cast this type to float and we are then doing comparisons based on the other type used, see https://cppinsights.io/s/1c62be16. It's not quite clear to me where this is failing at the moment.
If implicit conversion isn't sufficient so that we actually need to declare comparison operators, it seems most sensible to me to provide a templated overload that cast the half type to `float` and let the compiler deal with other casts required.

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: With the current approach we are getting
```
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:903:52: error: invalid operands to binary expression ('int' and 'const Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>')
    actual_lhs(LT_S_H)   = static_cast<int>(h_lhs) < h_rhs;
                           ~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:377:7: note: in instantiation of member function 'Test::Functor_TestHalfOperators<Kokkos::View<double *, Kokkos::Experimental::SYCL>, Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>::operator()' requested here
      run_on_host(0);
      ^
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:980:50: note: in instantiation of member function 'Test::Functor_TestHalfOperators<Kokkos::View<double *, Kokkos::Experimental::SYCL>, Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>::Functor_TestHalfOperators' requested here
  Functor_TestHalfOperators<ViewType, half_type> f_device(h_lhs, h_rhs);
                                                 ^
/home/darndt/kokkos/core/unit_test/TestHalfOperators.hpp:1052:5: note: in instantiation of function template specialization 'Test::__test_half_operators<Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>>' requested here
    __test_half_operators<half_t>(h_lhs + cast_to_half(i + 1),
    ^
/soft/restricted/CNDA/updates/2023.05.15.001/oneapi/compiler/eng-20230614/compiler/linux/bin-llvm/../include/sycl/half_type.hpp:530:3: note: candidate function not viable: no known conversion from 'const Kokkos::Experimental::Impl::floating_point_wrapper<sycl::detail::half_impl::half>' to 'const half' for 2nd argument
  OP(<)
  ^
[...]
```
(but `actual_lhs(LT_H_S)   = h_lhs < static_cast<int>(h_rhs);` still works).

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:835]: I would suggest doing
```diff
diff --git a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
index 0b240c88e..b06787cf1 100644
--- a/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
+++ b/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
@@ -820,18 +820,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs < tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) < rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<impl_type>(lhs) < rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<(T lhs, floating_point_wrapper rhs) {
-    return lhs < T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs < static_cast<impl_type>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -841,18 +839,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs > tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) > rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) > rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>(T lhs, floating_point_wrapper rhs) {
-    return lhs > T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs > static_cast<float>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -862,18 +858,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs <= tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<=(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) <= rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<=(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) <= rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator<=(T lhs, floating_point_wrapper rhs) {
-    return lhs <= T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator<=(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs <= static_cast<float>(rhs);
   }
 
   KOKKOS_FUNCTION
@@ -883,18 +877,16 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs >= tmp_rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>=(floating_point_wrapper lhs, T rhs) {
-    return T(lhs) >= rhs;
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>=(const floating_point_wrapper& lhs, const T& rhs) {
+    return static_cast<float>(lhs) >= rhs;
   }
 
-  template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
-  operator>=(T lhs, floating_point_wrapper rhs) {
-    return lhs >= T(rhs);
+  template <typename T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool>
+  operator>=(const T& lhs, const floating_point_wrapper& rhs) {
+    return lhs >= static_cast<float>(rhs);
   }
 
   // Insertion and extraction operators
```

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: No, https://cppinsights.io/s/2155d1dd shows that we would compare doubles in that case (the half_type is first cast to `float` and then to `double`).

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: @masterleinad: OK, thank you for the detailed input. We currently downcast for all mixed precision expressions unless we are mixing with float or double operands, in which case, we upcast the half-type to float or double. https://godbolt.org/z/vaWhahrb1 shows that when mixing a float with a integral-type in binary arithmetic or comparison operators, this same behavior holds: the compiler implicitly downcasts the `long int lhs` to a float. This PR is intended to address a specific bug when mixing half with single and double precision types but may be getting a bit mixed up with some larger feature requests. It's not clear to me whether the compiler will handle upcasting and downcasting for floating point types and integral types involving half-precision operands without being explicit so I pushed this change:
https://github.com/kokkos/kokkos/blob/630a520aa78329b0d8981a8f47cab5ab54d79ce1/core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp#L842-L857.
to maintain the proper upcast / downcast behavior in our LT comparison operator.

What do you think? Also are you OK with addressing mixed precision involving integral types during a separate issue / PR?

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: > [godbolt.org/z/vaWhahrb1](https://godbolt.org/z/vaWhahrb1) shows that when mixing a float with a integral-type in binary arithmetic or comparison operators, this same behavior holds: the compiler implicitly downcasts the long int lhs to a float. [...] It's not clear to me whether the compiler will handle upcasting and downcasting for floating point types and integral types involving half-precision operands without being explicit  [...]

Right, and that's the same behavior https://cppinsights.io/s/1cc93aec(slight variation of https://cppinsights.io/s/2155d1dd) shows: An `int` is cast to `float`, for a `double` input we cast the `half_type` to `double` and for a `float` input we don't need extra casts (apart from the one from `half_type` to `float` that is explicit).
This is consistent with https://godbolt.org/z/7se48K5df resp. https://cppinsights.io/s/6e31f3e4 (a slight variation of your example using `double` instead of `int`): We cast the `float` variable to `double`.

Casting integral types to `half_type` looks OKish but I would rather leave selecting the appropriate casts in the comparison operators to the compiler. (although checking for `!std::is_floating_point_v` introduces a comparison operator for arbitrary types like `std::string` which might not what we want).
Can you explain why you don't trust the compiler to do the "right thing" after converting the `half_type` to `float` (which doesn't lose information)?

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: It's not so much about not trusting the compiler to do the "right thing" as much as it is about ensuring we provide these operations in half-precision rather than relying on single precision arithmetic -- In short, it's not clear to me how the compiler can figure out that half-precision operations should be used without being explicit here.

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: > In short, it's not clear to me how the compiler can figure out that half-precision operations should be used without being explicit here.

Right, we wouldn't use `half-precision` operations if we cast to `float` explicitly and then it's a question about correctness vs. performance. 
I guess it would be best to forward the operation to the actual `half_type` in case that it's not just a storage type.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: Are you OK with merging this bug-fix and addressing integral types in a follow-on?

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: I think it's a pretty easy fix to do
```C++
KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float>, bool> 
operator<(const floating_point_wrapper& lhs, const T& rhs) { 
#ifdef KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
  return lhs.val < rhs;
#else
  return static_cast<float>(lhs) < rhs;
#endif
}
```
to actually use half-precision operations if possible. If you want to restrict this overload to `float` and `double` for now and add testing for `int` in a later pull request, that's fine with me.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: Are you suggesting we remove the member functions for LT,GT,LE, and GE?

Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:854]: Can we? Since we implement both `operator<(T, half_type)` and `operator<half_type, T`, I would expect `operator<(half_type, half_type)` to be ambiguous if not explicitly provided.

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 4]
```diff
@@ -903,15 +931,15 @@ class alignas(FloatType) floating_point_wrapper {
...
     return T(lhs) >= rhs;
   }
 
   template <class T>
-  KOKKOS_FUNCTION friend std::enable_if_t<
-      std::is_same_v<T, float> || std::is_same_v<T, double>, bool>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float> &&
+      (std::is_same_v<T, float> || std::is_same_v<T, double>), bool>
   operator>=(T lhs, floating_point_wrapper rhs) {
     return lhs >= T(rhs);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp]: What about these?

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp]: Good catch

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 5]
```diff
@@ -839,27 +839,139 @@ class alignas(FloatType) floating_point_wrapper {
...
+
+  template <class T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float> &&
+                                              (std::is_same_v<T, float> ||
+                                               std::is_same_v<T, double>),
+                                          bool>
+  operator>(T lhs, floating_point_wrapper rhs) {
+#ifdef KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
+    // CAUTION: this may result in different implicit casting across backends
+    return lhs.val > rhs;
```
Comment (masterleinad) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:899]:
```suggestion
    return lhs > rhs.val;
```

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp [hunk 6]
```diff
@@ -839,27 +839,139 @@ class alignas(FloatType) floating_point_wrapper {
...
+  template <class T>
+  KOKKOS_FUNCTION friend std::enable_if_t<std::is_convertible_v<T, float> &&
+                                              (std::is_same_v<T, float> ||
+                                               std::is_same_v<T, double>),
+                                          bool>
+  operator<(floating_point_wrapper lhs, T rhs) {
+#ifdef KOKKOS_HALF_IS_FULL_TYPE_ON_ARCH
+    // CAUTION: this may result in different implicit casting across backends
+    return lhs.val < rhs;
+#else
```
Comment (masterleinad) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:851]: Looks like for `Cuda` there is only the `operator<(__half, __half)` overload.

Comment (e10harvey) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:851]: Yes

**Review by masterleinad — APPROVED** [review]
